include/oneapi/dpl/internal/dynamic_selection_impl/policy_base.h

```diff
@@ -0,0 +1,155 @@
+    }
+
+    void
+    initialize()
+    {
+        if (!backend_)
+            backend_ = std::make_shared<backend_t>(ReportReqs{}...);
+        static_cast<Policy*>(this)->initialize_impl();
+    }
+
+    template <typename Arg0, typename... Args,
+              typename = std::enable_if_t<!std::is_same_v<std::decay_t<Arg0>, std::vector<resource_type>>>>
+    void
+    initialize(Arg0&& arg0, Args&&... args)
+    {
+        if (!backend_)
+            backend_ = std::make_shared<backend_t>(ReportReqs{}...);
+        static_cast<Policy*>(this)->initialize_impl(std::forward<Arg0>(arg0), std::forward<Args>(args)...);
+    }
+
+    void
+    initialize(const std::vector<resource_type>& u)
+    {
+        if (!backend_)
+            backend_ = std::make_shared<backend_t>(u, oneapi::dpl::identity(), ReportReqs{}...);
+        static_cast<Policy*>(this)->initialize_impl();
+    }
+
+    template <typename Arg0, typename... Args,
+              typename = std::enable_if_t<!std::is_same_v<std::decay_t<Arg0>, ResourceAdapter>>>
+    void
+    initialize(const std::vector<resource_type>& u, Arg0&& arg0, Args&&... args)
+    {
+        if (!backend_)
+            backend_ = std::make_shared<backend_t>(u, oneapi::dpl::identity(), ReportReqs{}...);
+        static_cast<Policy*>(this)->initialize_impl(std::forward<Arg0>(arg0), std::forward<Args>(args)...);
+    }
+
+    template <typename... Args>
+    void
+    initialize(const std::vector<resource_type>& u, ResourceAdapter adapter, Args... args)
+    {
+        if (!backend_)
+            backend_ = std::make_shared<backend_t>(u, adapter, ReportReqs{}...);
+        static_cast<Policy*>(this)->initialize_impl(args...);
+    }
+
+    template <typename Function, typename... Args>
+    auto //std::shared_ptr of the "wait type"
+    try_submit(Function&& f, Args&&... args)
+    {
+        using ret_t =
+            decltype(std::declval<Backend>().submit(std::declval<decltype(*std::declval<Policy>().try_select_impl(
+                                                        std::declval<Function>(), std::declval<Args>()...))>(),
+                                                    std::forward<Function>(f), std::forward<Args>(args)...));
+        if (backend_)
+        {
+            auto e = static_cast<Policy*>(this)->try_select_impl(f, args...);
+            if (!e)
+            {
+                // return an empty shared_ptr
+                return std::shared_ptr<ret_t>{};
+            }
+            else
+            {
+                return std::make_shared<ret_t>(
+                    backend_->submit(*e, std::forward<Function>(f), std::forward<Args>(args)...));
+            }
+        }
+        throw std::logic_error("submit called before initialization");
+    }
+
+    template <typename Function, typename... Args>
+    auto
+    submit(Function&& f, Args&&... args)
+    {
+        return oneapi::dpl::experimental::internal::submit_fallback(
```

Comment:
Will this correctly find a submit_impl if defined in the child class?  Similar comment for submit_and_wait.